src/lsm/block_free_set.zig

```diff
@@ -1,37 +1,421 @@
+        // Update the index when every block in the shard is allocated.
+        if (set.find_free_block_in_shard(shard) == null) set.index.unset(shard);
+
+        const address = block + 1;
         return @intCast(u64, address);
     }
 
+    fn find_free_block_in_shard(set: *BlockFreeSet, shard: usize) ?usize {
+        const shard_start = shard * shard_size;
+        const shard_end = shard_start + shard_size;
+        assert(shard_start < set.blocks.bit_length);
+
+        return find_first_set_bit(set.blocks, shard_start, shard_end);
+    }
+
+    fn is_free(set: *BlockFreeSet, address: u64) bool {
+        const block = address - 1;
+        return set.blocks.isSet(block);
+    }
+
+    /// Given the address, marks an allocated block as free.
     pub fn release(set: *BlockFreeSet, address: u64) void {
-        const bit = address - 1;
-        assert(!set.free.isSet(bit));
-        set.free.set(bit);
+        const block = address - 1;
+        assert(!set.blocks.isSet(block));
+
+        set.index.set(block / shard_size);
+        set.blocks.set(block);
+        set.recent.push(block) catch {}; // Ignore error.NoSpaceLeft if full.
+    }
+
+    /// Decodes the compressed bitset in `source` into `set`. Panics if `source`'s encoding is invalid.
+    pub fn decode(set: *BlockFreeSet, source: []align(@alignOf(usize)) const u8) void {
+        // Verify that this BlockFreeSet is entirely unallocated.
+        assert(set.index.count() == set.index.bit_length);
+
+        const words_decoded = ewah.decode(source, bitset_masks(set.blocks));
+        assert(words_decoded * @bitSizeOf(MaskInt) <= set.blocks.bit_length);
+
+        var shard: usize = 0;
+        while (shard < set.index.bit_length) : (shard += 1) {
+            if (set.find_free_block_in_shard(shard) == null) set.index.unset(shard);
+        }
+    }
+
+    // Returns the maximum number of bytes that the `BlockFreeSet` needs to encode to.
+    pub fn encode_size_max(set: BlockFreeSet) usize {
+        return ewah.encode_size_max(bitset_masks(set.blocks));
+    }
+
+    // Returns the number of bytes written to `target`.
+    pub fn encode(set: BlockFreeSet, target: []align(@alignOf(usize)) u8) usize {
+        assert(target.len == set.encode_size_max());
+
+        return ewah.encode(bitset_masks(set.blocks), target);
     }
 };
+
+fn bitset_masks(bitset: DynamicBitSetUnmanaged) []usize {
+    const len = div_ceil(bitset.bit_length, @bitSizeOf(MaskInt));
+    return bitset.masks[0..len];
+}
+
+test "BlockFreeSet acquire/release" {
+    const block_size = config.lsm_table_block_size;
+    const blocks_in_tb = (1 << 40) / block_size;
+    try test_block_shards_count(5120 * 8, 10 * blocks_in_tb);
+    try test_block_shards_count(5120 * 8 - 1, 10 * blocks_in_tb - BlockFreeSet.shard_size);
+    try test_block_shards_count(1, BlockFreeSet.shard_size); // At least one index bit is required.
+    // Block counts are not necessarily a multiple of the word size.
```

Comment:
We should rather require that they are, which I think we're doing now? This might just be an older comment?

Reply:
You're right, that's just an outdated comment.